tutorial/spiders/pakistan_stores_spider.py

```diff
@@ -0,0 +1,97 @@
+"""
+This module contains a spider of pakistanistores.com site
+that crawls and extracts data
+"""
+import json
+from functools import partial
+
+import scrapy
+
+
+class PakistanStoresSpider(scrapy.Spider):
+    """
+    This class crawls the site pakistanistores.com and extracts data
+    """
+
+    name = "pakistan_stores"
+    current_page = 1
+    products = []
+    processed_products = 0
+
+    def start_requests(self):
+        url = getattr(self, "url", None)
+        yield scrapy.Request(url, self.parse)
+
+    def parse(self, response, **kwargs):
+        """
+        This function crawls through each page and extracts all product information
+        Args:
+            response(TextReponse): Contains the page contents
+        Returns:
+            (Any or None):  A request to another page
+                            Or
+                            None if everything processed
+        """
+        last_page = response.css("a.page-link.navigate::attr(data-href)")[-1].get()
+        last_page_number = int(last_page[last_page.find("=") + 1 :])
+        product_containers = response.css("li.col-md-3.col-md-3.col-sm-6.col-xs-6 a")
+        for product_container in product_containers:
+            name = product_container.attrib["title"]
+            url = response.urljoin(product_container.attrib["href"])
+            img_url = product_container.css("img.lazyload").attrib["data-src"]
+            price = product_container.css("div.primary-color.price::text").get()[0:-1]
+            product = {
+                "name": name,
+                "link": url,
+                "img_link": response.urljoin(img_url),
+                "price": price,
+            }
+            self.products.append(product)
+        self.current_page += 1
+        if self.current_page <= last_page_number:
+            next_page_url = response.urljoin(f"?page={self.current_page}")
+            yield scrapy.Request(url=next_page_url, callback=self.parse)
+        else:
+            for product in self.products:
+                link = product["link"]
+                binded_f = partial(self.extract_description, product)
+                product["description"] = ""
+                if link.__contains__("/product/"):
+                    self.increment_and_check()
+                else:
+                    yield scrapy.Request(url=link, callback=binded_f)
+
+    def extract_description(self, product, response):
+        """
+        This function extracts description of the product from the response
+        Args:
+            product(dict):  Dictionary containing product information
+                            like, name, link, img_link and price
+            response(TextResponse): Response object containing page contents
+        Returns:
+            None
+        """
+        index = self.products.index(product)
+        description = response.css("div.light p::text")[-1].get()[1:-1]
```

Comment:
How about if we want to make something reusable here? Can we define `::text` prop in some common place and everytime we use it from there?

Reply:
I have created a file `constants.py` and defined a Constant `TEXT_PROP` and used it where i was previously using `::text`